### PR TITLE
[INFRANG-6172] Add some code to perform ddb stream cutover

### PR DESCRIPTION
**Pull request by andruwm**
# Jira
https://clever.atlassian.net/browse/INFRANG-6172

# Overview
This PR needs to be deployed before #77

This will work like so:
1. us-west-1 will be deployed with this change and continue to process records until the cutover time
2. #77 will be deployed and us-west-1 will be orphaned, which will continue to plug along until cutover.
3. The new region will start processing after the cutover
4. Once cutover happens, delete the old region then revert this PR.

# Testing
I will deploy this PR to clever-dev first for testing.

### CODE REVIEW

File: cmd/dynamodb/main.go [hunk 1]
```diff
@@ -84,6 +88,40 @@ func main() {
...
+// Facilitates cutting over to a new DDB stream. Before the time all
+// records will be processed in uw1. After the time, all records will be
+// processed in uw2. Making the assumption that it is extremely unlikely
+// that a record will have an exact timestamp down to the nanosecond of
+// the cutover time, so skipping that edge case. If you are reading this
+// after the cutover time you can safely revert
+// https://github.com/Clever/ddb-to-es/pull/78
+func skipRecord(record events.DynamoDBEventRecord) (bool, error) {
+	// only do this in prod and clever-dev
+	env := os.Getenv("_DEPLOY_ENV")
```
Comment (taylor-sutton) [inline cmd/dynamodb/main.go:102]:
```suggestion
	env := os.Getenv("DEPLOY_ENV")
```

AWS doesn't allow lambdas to have env vars beginning with underscore, so we don't use it. See e.g https://us-west-1.console.aws.amazon.com/cloudformation/home?region=us-west-1#/stacks/template?filteringText=ddb-to-es&filteringStatus=active&viewNested=true&stackId=arn%3Aaws%3Acloudformation%3Aus-west-1%3A589690932525%3Astack%2Fproduction--ddb-to-es%2F4729e840-1035-11e8-92d3-500cadc9fefe

File: cmd/dynamodb/main.go [hunk 2]
```diff
@@ -84,6 +88,40 @@ func main() {
...
+	if env != "production" && env != "clever-dev" {
+		return false, nil
+	}
+
+	t, err := time.Parse(time.RFC3339, cutoverTime)
+	if err != nil {
+		return false, fmt.Errorf("failed to parse cutover time: %s", err)
+	}
+	r, ok := os.LookupEnv("_POD_REGION")
```
Comment (taylor-sutton) [inline cmd/dynamodb/main.go:111]:
```suggestion
	r, ok := os.LookupEnv("POD_REGION")
```

File: cmd/dynamodb/main.go [hunk 3]
```diff
@@ -117,6 +162,10 @@ func processRecords(records []events.DynamoDBEventRecord, db es.DB) ([]es.Doc, e
 		}
 	}
 
+	if len(docs) == 0 {
+		return nil, errors.New("all records skipped for stream cutover")
```
Comment (taylor-sutton) [inline cmd/dynamodb/main.go:166]: Is this right to be returned as an error? It seems like it successfully did nothing - for example we don't want Lambda to retry.

**Review by taylor-sutton — COMMENTED** [review]
Couple of small things. Also when you're testing this, I'd be curious if having a us-west-2 lambda have events from a us-west-1 DDB stream works out-of-the-box, or if you also will need an ark-config change to switch to the us-west-2 stream.

File: cmd/dynamodb/main.go (hunk 1 above)
Comment (andruwm) [reply cmd/dynamodb/main.go:102]: Oh, that explains why this wasnt working in my testing on Friday. I'll make this change then test again today before deploying.

**Comment by andruwm** [comment]
> Couple of small things. Also when you're testing this, I'd be curious if having a us-west-2 lambda have events from a us-west-1 DDB stream works out-of-the-box, or if you also will need an ark-config change to switch to the us-west-2 stream.

#77 will get deployed after this PR, it has a corresponding ark-config PR

File: cmd/dynamodb/main.go (hunk 3 above)
Comment (andruwm) [reply cmd/dynamodb/main.go:183]: The lambda only logs errors and doesn't return them as it is configured currently. Seeing as this is temporary, it makes it easy to see if the change actually worked so I am gonna leave it as is.

**Review by taylor-sutton — APPROVED** [review]

**Pull request merged**